pilot/pkg/xds/endpoints/endpoint_builder.go

```diff
@@ -443,9 +415,82 @@ func (b *EndpointBuilder) BuildLocalityLbEndpointsFromShards(
 		b.push.AddMetric(model.ProxyStatusClusterNoInstances, b.clusterName, "", "")
 	}
 
+	// Apply the Split Horizon EDS filter, if applicable.
+	locEps = b.EndpointsByNetworkFilter(locEps)
+
+	if model.IsDNSSrvSubsetKey(b.clusterName) {
+		// For the SNI-DNAT clusters, we are using AUTO_PASSTHROUGH gateway. AUTO_PASSTHROUGH is intended
+		// to passthrough mTLS requests. However, at the gateway we do not actually have any way to tell if the
+		// request is a valid mTLS request or not, since its passthrough TLS.
+		// To ensure we allow traffic only to mTLS endpoints, we filter out non-mTLS endpoints for these cluster types.
+		locEps = b.EndpointsWithMTLSFilter(locEps)
+	}
+
 	return locEps
 }
 
+func (b *EndpointBuilder) filterIstioEndpoint(ep *model.IstioEndpoint, svcPort *model.Port) bool {
```

Comment:
This needs to be looked at closely. Must cover what we have in the original CDS code.